bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/auth/client/oauth2/StorageCipher.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.auth.client.oauth2;
+
+import java.security.GeneralSecurityException;
+
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * This allows the encryption and decryption to be performed before saving to storage.
+ *
+ * @author Gary Tse - initial contribution
+ *
+ */
+public interface StorageCipher {
+
+    /**
+     * A unique cipher identifier per each implementation of StorageCipher.
+     * It allows the OAuthStoreHandler to choose which cipher implementation to use.
+     * This is particularly important when old ciphers becomes out-dated and
+     * need to be replaced by new implementations.
+     *
+     * @return unique identifier
+     */
+    public String getUniqueCipherId();
+
+    /**
+     * Encrypt the plainText, then produce a base64 encoded cipher text
+     *
+     * @param plainText
+     * @return base64-encoded( encrypted( text ) )
+     * @throws GeneralSecurityException
```

Comment:
Shouldn't the exception be described when it would be thrown?

Reply:
Similar to above

Reply:
Another example would be `SecurityException` which is also documented clearly in javadoc.

```javadoc
/**
     * Creates a <code>FileInputStream</code> by
     * opening a connection to an actual file,
     * the file named by the path name <code>name</code>
     * in the file system.  A new <code>FileDescriptor</code>
     * object is created to represent this file
     * connection.
     * <p>
     * First, if there is a security
     * manager, its <code>checkRead</code> method
     * is called with the <code>name</code> argument
     * as its argument.
     * <p>
     * If the named file does not exist, is a directory rather than a regular
     * file, or for some other reason cannot be opened for reading then a
     * <code>FileNotFoundException</code> is thrown.
     *
     * @param      name   the system-dependent file name.
     * @exception  FileNotFoundException  if the file does not exist,
     *                   is a directory rather than a regular file,
     *                   or for some other reason cannot be opened for
     *                   reading.
     * @exception  SecurityException      if a security manager exists and its
     *               <code>checkRead</code> method denies read access
     *               to the file.
     * @see        java.lang.SecurityManager#checkRead(java.lang.String)
     */
    public FileInputStream(String name) throws FileNotFoundException {
        this(name != null ? new File(name) : null);
    }
```

I personally would prefer all exceptions to be included in javadoc with proper descriptions since it is a general practice.